Firestore/core/test/unit/remote/bloom_filter_test.cc

```diff
@@ -132,6 +143,153 @@ TEST(BloomFilterTest,
   }
 }
 
+class BloomFilterGoldenTest : public ::testing::Test {
+ public:
+  void RunGoldenTest(const std::string& test_file) {
+    BloomFilter bloom_filter = GetBloomFilter(test_file);
+    std::string membership_result = GetMembershipResult(test_file);
+
+    for (size_t i = 0; i < membership_result.length(); i++) {
+      bool expectedResult = membership_result[i] == '1';
+      bool mightContainResult =
+          bloom_filter.MightContain(kGoldenDocumentPrefix + std::to_string(i));
+
+      EXPECT_EQ(mightContainResult, expectedResult);
+    }
+  }
+
+ private:
+  static const char* kGoldenDocumentPrefix;
+  JsonReader reader;
+
+  static Path GetGoldenTestFolder() {
+    return Path::FromUtf8(__FILE__).Dirname().AppendUtf8(
+        "bloom_filter_golden_test_data/");
+  }
+
+  json ReadFile(const std::string& file_name) {
+    Path file_path = GetGoldenTestFolder().AppendUtf8(file_name);
+    std::ifstream stream(file_path.native_value());
+    HARD_ASSERT(stream.good());
+    return nlohmann::json::parse(stream);
+  }
+
+  BloomFilter GetBloomFilter(const std::string& file_name) {
+    json test_file = ReadFile(file_name);
+    nlohmann::json bits = reader.OptionalObject("bits", test_file, {});
+    std::string bitmap = reader.OptionalString("bitmap", bits, "");
+    int padding = reader.OptionalInt("padding", bits, 0);
+    int hash_count = reader.OptionalInt("hashCount", test_file, 0);
+    std::string decoded;
+    absl::Base64Unescape(bitmap, &decoded);
+    std::vector<uint8_t> decoded_map(decoded.begin(), decoded.end());
+
+    StatusOr<BloomFilter> maybe_bloom_filter =
+        BloomFilter::Create(std::move(decoded_map), padding, hash_count);
+    HARD_ASSERT(maybe_bloom_filter.ok(),
+                "Bloom filter input file %s has invalid values.", file_name);
+    BloomFilter bloom_filter = maybe_bloom_filter.ValueOrDie();
+
+    return bloom_filter;
+  }
+
+  std::string LocateResultFile(std::string file_name) {
+    const std::string substring = "bloom_filter_proto";
+    size_t start_pos = file_name.find(substring);
+    HARD_ASSERT(start_pos != std::string::npos,
+                "Test file name %s is not valid, expected to include "
+                "\"bloom_filter_proto\".",
```

Comment:
Don't copy the string "bloom_filter_proto" here. Instead, use "%s" in the message and `substring` as an argument to the HARD_ASSERT.